mlx/c/random.cpp

```diff
@@ -4,199 +4,332 @@
-      loc,
-      scale,
-      (key ? std::make_optional(key->ctx) : std::nullopt),
-      s->ctx));
-}
-extern "C" mlx_array mlx_random_randint(
-    mlx_array low,
-    mlx_array high,
+    const mlx_array key /* may be null */,
+    const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res,
+        mlx::core::random::normal(
+            std::vector<int>(shape, shape + shape_num),
+            mlx_dtype_to_cpp(dtype),
+            loc,
+            scale,
+            (key.ctx ? std::make_optional(mlx_array_get_(key)) : std::nullopt),
+            mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
+}
+extern "C" int mlx_random_randint(
+    mlx_array* res,
+    const mlx_array low,
+    const mlx_array high,
     const int* shape,
-    size_t num_shape,
-    mlx_array_dtype dtype,
-    mlx_array key,
-    mlx_stream s) {
-  RETURN_MLX_C_ARRAY(mlx::core::random::randint(
-      low->ctx,
-      high->ctx,
-      MLX_CPP_INTVEC(shape, num_shape),
-      MLX_CPP_ARRAY_DTYPE(dtype),
-      (key ? std::make_optional(key->ctx) : std::nullopt),
-      s->ctx));
-}
-extern "C" void mlx_random_seed(uint64_t seed) {
-  RETURN_MLX_C_VOID(mlx::core::random::seed(seed));
-}
-extern "C" mlx_array
-mlx_random_split_equal_parts(mlx_array key, int num, mlx_stream s) {
-  RETURN_MLX_C_ARRAY(mlx::core::random::split(key->ctx, num, s->ctx));
-}
-extern "C" mlx_tuple_array_array mlx_random_split(mlx_array key, mlx_stream s) {
-  RETURN_MLX_C_ARRAYPAIR(mlx::core::random::split(key->ctx, s->ctx));
-}
-extern "C" mlx_array mlx_random_truncated_normal(
-    mlx_array lower,
-    mlx_array upper,
+    size_t shape_num,
+    mlx_dtype dtype,
+    const mlx_array key /* may be null */,
+    const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res,
+        mlx::core::random::randint(
+            mlx_array_get_(low),
+            mlx_array_get_(high),
+            std::vector<int>(shape, shape + shape_num),
+            mlx_dtype_to_cpp(dtype),
+            (key.ctx ? std::make_optional(mlx_array_get_(key)) : std::nullopt),
+            mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
+}
+extern "C" int mlx_random_seed(uint64_t seed) {
+  try {
+    mlx::core::random::seed(seed);
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
+}
+extern "C" int mlx_random_split_equal_parts(
+    mlx_array* res,
+    const mlx_array key,
+    int num,
+    const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res,
+        mlx::core::random::split(mlx_array_get_(key), num, mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
+}
+extern "C" int mlx_random_split(
+    mlx_array* res_0,
+    mlx_array* res_1,
+    const mlx_array key,
+    const mlx_stream s) {
+  try {
+    std::tie(mlx_array_get_(*res_0), mlx_array_get_(*res_1)) =
```

Comment:
Another case of:

```
MLX error: expected a non-empty mlx_array at mlx-swift2/Source/Cmlx/include/mlx/c/random.cpp:281
```

I think `mlx_array_get()` is not meant to be used on out parameters.

Reply:
fixed, thanks!